### ISSUE TITLE
.NET 5 Causes Unhandled ArgumentNullException on WM_LBUTTONDOWN

### ISSUE THREAD

**Issue by StevenBonePgh** [FizzcodeSoftware/ObjectListViewRepack#1]
If your application built/running under .NET 5 crashes when you click the first column in details view, with an ArgumentNullException for parameter 'owningItem' that is raised with this stack trace:

```
   at System.Windows.Forms.ListViewItem.ListViewSubItem.ListViewSubItemAccessibleObject..ctor(ListViewSubItem owningSubItem, ListViewItem owningItem)
   at System.Windows.Forms.ListView.ListViewAccessibleObject.HitTest(Int32 x, Int32 y)
   at System.Windows.Forms.ListView.WmMouseDown(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.ListView.WndProc(Message& m)
   at ObjectListView.WndProc(Message& m) in 
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
```
This appears to be caused by a regression in .NET 5's ListView. I reported this to the Winform team here: https://github.com/dotnet/winforms/issues/4742

**Comment by rsl-outlook** [comment]
Hi,

Great catch.

I notice that it's been fixed but how do I go about getting the fix?

**Comment by StevenBonePgh** [comment]
Next monthly .NET 5 update is the best I can figure when it will be released, so mid May 2021. Getting the fix would mean the runtime being updated (so it may need an installer update if you are distributing the runtime with the application), or if your executable is being packaged WITH the runtime you'd need to rebuild the application with the new SDK installed, or so I think. This framework servicing model and its implications is pretty new to me. Rebuilding the library would have no impact, as far as I understand. 

Another thing I noticed is that you probably also want to multitarget this project and provide a .NET 5 specific variant to prevent a different issue under .NET 5, since .NET 5 now exposes `GroupImageList `in `ListView`. This means anyone using this property via the library would be effectively treating this as if a 'new' keyword was used for it - meaning any code that treats it as a `ListView` type will get strange behaviors with the `GroupImageList`. The fix would be something like this:
```
#if NET5_0_OR_GREATER
        // GroupImageList is part of .NET 5 natively.
#else
        /// <summary>
        /// Gets or sets the image list from which group header will take their images
        /// </summary>
        /// <remarks>If this is not set, then group headers will not show any images.</remarks>
        [Category("ObjectListView"),
         Description("The image list from which group header will take their images"),
         DefaultValue(null)]
        public ImageList GroupImageList {
            get { return this.groupImageList; }
            set {
                this.groupImageList = value;
                if (this.Created) {
                    NativeMethods.SetGroupImageList(this, value);
                }
            }
        }
        private ImageList groupImageList;
#endif
```
To multitarget, I'd suggest perhaps changing `ObjectListView2019Core3.csproj` like this:
`<TargetFramework>netcoreapp3.1</TargetFramework>`
to
`<TargetFrameworks>net4;netcoreapp3.1;net5.0-windows</TargetFrameworks>`
Then one package can properly support classic Framework and fully support two .NET core targets.

**Comment by rsl-outlook** [comment]
I tried to install the .net 5.0.202 (version 5.0.5) build from 6th April but it didn't seem to make any difference. From what you've said it seems that this fix doesn't form part of that release anyway. You're quite right that nothing can be done with this library as it's a problem with the underlying framework listview control.

Thanks for also raising the point regarding the GroupImageList. Multi targeting for my app does not play nicely with another library that it consumes so I've got some hard decisions to make. Luckily it's just a demo so I can change the controls to something else for now.

Thanks again

**Comment by StevenBonePgh** [comment]
I see the fix should be in [5.0.6](https://github.com/dotnet/winforms/milestone/41?closed=1) and [6.0 Preview 4](https://github.com/dotnet/winforms/milestone/40?closed=1). Look for **Fix ANE in the "ListViewSubItemAccessibleObjectin " constructor**

There are only a few issues I have encountered from a project consuming another project that is multitargeted. First, and this is unlikely to be the case if you are using SDK style projects as the consumer (which I am sure you are since this is targeted only for .NET Core), is that some project types that are ignorant of the new project system only will use the first target framework from the list of target frameworks. This is a problem for current version Xamarin which would require you to place NetStandard first, for example.

Given that constraint, there are two issues I have reported with Visual Studio that could make it appear like you have an issue. The workaround for either of these appears to be referencing the nuget package rather than the project. I'll link the bugs below, and would appreciate any votes on them as well, as they are significant hurdles when one is attempting to migrate to .NET Core from .NET Framework.

[This bug is that Visual Studio, specifically the Winforms Designer, references the first target rather than the appropriate target - the one it uses to reference when compiling.](https://developercommunity.visualstudio.com/t/winforms-designer-uses-assembly-from-wrong-target/1379457)

[This other bug is also designer related where .NET Core projects will fail to load in the designer in some cases.](https://developercommunity.visualstudio.com/t/net-5-winforms-designer-load-error-does-not-load-p/1384393) The specific scenario I listed is merely what I can reproduce easily, but I think you will experience this at times when referencing any winforms base class that comes from another project. The workaround, if you experience it, is to compile the solution, close Visual Studio, and reload the solution. The designer will then work for a while, until whatever trigger happens that causes it to fail again - I have not been able to nail down a repro for that. This one is NOT specific to multitargeted projects, but the intermittent nature of it could possibly make you think that it does by coincidence.

**Comment by rsl-outlook** [comment]
I'll look out for 5.0.6.

Have voted up the two issues that you linked to.

Unfortunately I get the impression that development effort is going into new MAUI and the like rather than supporting this more mature technology.

By the way - I also find I can't edit the columns property of an ObjectListView correcrtly in Visual studio. Instead of giving me the OLV columns dialog it shows the standard ListView columns dialog and so I have to create OlvColumns in code instead. Not sure if this is a bug or just an expected consequence of the port over to net core?

This is what happens when I try and add a column from the designer:

![image](https://user-images.githubusercontent.com/80892367/114675318-0129a780-9d00-11eb-81c2-d86f35b4ac02.png)

**Comment by StevenBonePgh** [comment]
This is yet another bug, or inconsistency, in the Visual Studio WinForms designer for .NET Core vs the .NET Framework. I did a bit of research and [submitted a Feedback Item for it](https://developercommunity.visualstudio.com/t/net-5-winforms-designer-ignores-systemcomponentmod/1398197). This one will probably need votes as well. 

Best guess as to the cause is the use of the 'new' keyword:
```
[System.ComponentModel.Editor("BrightIdeasSoftware.Design.OLVColumnCollectionEditor", "System.Drawing.Design.UITypeEditor")]
public new ListView.ColumnHeaderCollection Columns
```
My money is on the .NET Core designer's reflection on the control does not pick up our `EditorAttribute` on the 'new' property like the .NET Framework's editor did. I've used this approach a handful of times to get custom editors on top of pre-existing properties, so I hope they fix it (normally I call to base, but for the OLV to work Phillip needs to use derived types in the collection).
